File: ios/ios/ViewModels/JobDetailViewModel.swift
//
//  JobDetailViewModel.swift
//  ios
//
//  Job detail screen view model
//

import Foundation
import SwiftUI

@MainActor
class JobDetailViewModel: ObservableObject {
    @Published var jobDetail: JobDetailResponse?
    @Published var builds: [Build] = []
    @Published var isLoading = false
    @Published var isRefreshing = false
    @Published var errorMessage: String?
    @Published var showError = false
    @Published var triggerMessage: String?
    @Published var showTriggerAlert = false
    @Published var showParametersSheet = false
    
    private let api = JenkinsAPI.shared
    let jobName: String
    let jobURL: String
    
    init(jobName: String, jobURL: String) {
        self.jobName = jobName
        self.jobURL = jobURL
    }
    
    /// Check if the job has parameters
    var hasParameters: Bool {
        jobDetail?.hasParameters ?? false
    }
    
    /// Get parameter definitions
    var parameters: [ParameterDefinition] {
        jobDetail?.parameterDefinitions ?? []
    }
    
    func loadData() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil
        
        do {
            let detail = try await api.fetchJobDetailByURL(jobURL: jobURL)
            jobDetail = detail
            builds = detail.builds ?? []
        } catch let error as APIError {
            errorMessage = error.localizedDescription
            showError = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }
        
        isLoading = false
    }
    
    func refresh() async {
        isRefreshing = true
        
        do {
            let detail = try await api.fetchJobDetailByURL(jobURL: jobURL)
            jobDetail = detail
            builds = detail.builds ?? []
        } catch {
            // Silently fail on refresh
        }
        
        isRefreshing = false
    }
    
    /// Called when user taps the trigger build button
    func onTriggerBuildTapped() {
        if hasParameters {
            // Show parameters sheet
            showParametersSheet = true
        } else {
            // No parameters, trigger directly
            triggerBuild(with: nil)
        }
    }
    
    /// Trigger build with optional parameters
    func triggerBuild(with parameters: [String: String]?) {
        Task {
            do {
                try await api.triggerBuildByURL(jobURL: jobURL, parameters: parameters)
                triggerMessage = "已触发构建"
                showTriggerAlert = true
                
                // Refresh after a short delay
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                await refresh()
            } catch let error as APIError {
                triggerMessage = "触发失败: \(error.localizedDescription)"
                showTriggerAlert = true
            } catch {
                triggerMessage = "触发失败: \(error.localizedDescription)"
                showTriggerAlert = true
            }
        }
    }
}

File: android/app/src/main/java/com/by/android/ui/jobdetail/JobDetailViewModel.kt
package com.by.android.ui.jobdetail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.by.android.data.model.Build
import com.by.android.data.model.JobDetailResponse
import com.by.android.data.model.ParameterDefinition
import com.by.android.data.repository.ApiResult
import com.by.android.data.repository.JenkinsRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class JobDetailUiState(
    val jobDetail: JobDetailResponse? = null,
    val builds: List<Build> = emptyList(),
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val errorMessage: String? = null,
    val triggerMessage: String? = null,
    val showParametersDialog: Boolean = false
) {
    /** Check if this job has parameters */
    val hasParameters: Boolean
        get() = jobDetail?.hasParameters == true
    
    /** Get parameter definitions */
    val parameters: List<ParameterDefinition>
        get() = jobDetail?.parameterDefinitions ?: emptyList()
}

class JobDetailViewModel(
    private val repository: JenkinsRepository,
    val jobName: String,
    private val jobUrl: String
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(JobDetailUiState())
    val uiState: StateFlow<JobDetailUiState> = _uiState.asStateFlow()
    
    init {
        loadData()
    }
    
    fun loadData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            
            // Use URL-based method to support jobs under view paths
            when (val result = repository.getJobDetailByUrl(jobUrl)) {
                is ApiResult.Success -> {
                    _uiState.update { 
                        it.copy(
                            jobDetail = result.data,
                            builds = result.data.builds ?: emptyList(),
                            isLoading = false
                        ) 
                    }
                }
                is ApiResult.Error -> {
                    _uiState.update { 
                        it.copy(isLoading = false, errorMessage = result.message) 
                    }
                }
            }
        }
    }
    
    fun refresh() {
        viewModelScope.launch {
            _uiState.update { it.copy(isRefreshing = true) }
            
            when (val result = repository.getJobDetailByUrl(jobUrl)) {
                is ApiResult.Success -> {
                    _uiState.update { 
                        it.copy(
                            jobDetail = result.data,
                            builds = result.data.builds ?: emptyList(),
                            isRefreshing = false
                        ) 
                    }
                }
                is ApiResult.Error -> {
                    _uiState.update { it.copy(isRefreshing = false) }
                }
            }
        }
    }
    
    /** Called when user taps the trigger build button */
    fun onTriggerBuildTapped() {
        if (_uiState.value.hasParameters) {
            // Show parameters dialog
            _uiState.update { it.copy(showParametersDialog = true) }
        } else {
            // No parameters, trigger directly
            triggerBuild(null)
        }
    }
    
    /** Hide the parameters dialog */
    fun hideParametersDialog() {
        _uiState.update { it.copy(showParametersDialog = false) }
    }
    
    /** Trigger build with optional parameters */
    fun triggerBuild(parameters: Map<String, String>?) {
        viewModelScope.launch {
            // Hide dialog first
            _uiState.update { it.copy(showParametersDialog = false) }
            
            when (val result = repository.triggerBuildByUrl(jobUrl, parameters)) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(triggerMessage = "已触发构建") }
                    kotlinx.coroutines.delay(1000)
                    refresh()
                }
                is ApiResult.Error -> {
                    _uiState.update { 
                        it.copy(triggerMessage = "触发失败: ${result.message}") 
                    }
                }
            }
        }
    }
    
    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }
    
    fun clearTriggerMessage() {
        _uiState.update { it.copy(triggerMessage = null) }
    }
}

